xds/internal/client/client.go

```diff
@@ -312,6 +363,46 @@ var newAPIClient = func(apiVersion version.TransportAPI, cc *grpc.ClientConn, op
 	return cb.Build(cc, opts)
 }
 
+// LDSUpdateWithMD contains the LDSUpdate and metadata, including version, raw
```

Comment:
This could work. We will still need an exported struct that users can use, but it can just call the methods on xds_client.

I don't want to add more features to the xds_client. It's getting too big.

I think it would make more sense to split xds_client, and move the maps into a separate struct (`cache`?). Then the cache can implement those methods.
This would take more time to do, as it moves code around, and need to fix the tests.

Reply:
> I think it would make more sense to split xds_client, and move the maps into a separate struct (`cache`?). Then the cache can implement those methods.
> This would take more time to do, as it moves code around, and need to fix the tests.

This makes more sense, and was what came to my mind initially as well. Maybe at least add a TODO saying that we should do that at some point in time?

But I'm still not very convinced about the need for this intermediate type `LDSUpdateWithMD`. Why not build the exact response type as required by the CSDS? Or at the minimum, there is no need to send the whole `ListenerUpdate` struct to whoever is building the dump required by CSDS, right. All it needs it the raw proto and the metadata.

Reply:
> the exact response type

Did you mean the proto message? The message is quite complex (https://github.com/menghanl/grpc-go/blob/csds/xds/csds/csds.go#L132). I think it's better to keep it out of the xds_client.

> All it needs it the raw proto and the metadata

I changed the struct to have only `Raw`, not the whole `Update`.

Reply:
Given that we have a struct per resource (eg: `LDSUpdateWithMD`), why does this have to be an `anypb.Any`? Can we either make it a type specific to the resource, or can we just have one struct instead of four?

Reply:
We need `anypb.Any` because that's what the CSDS messages need.
Good point that we only need one struct now.